pkg/agent/handler/credential.go

```diff
@@ -69,83 +73,164 @@ func (h *credentials) Handle(ctx context.Context, meta *proto.EventMeta, resourc
 		return nil
 	}
 
-	if ok := h.shouldProcessPending(cr.Status, cr.Metadata.State); ok {
+	if ok := h.shouldProcessDeleting(cr); ok {
+		logger.Trace("processing resource in deleting state")
+		h.onDeleting(ctx, cr)
+		return nil
+	}
+
+	var credential *management.Credential
+	if ok := h.shouldProcessPending(cr); ok {
 		log.Trace("processing resource in pending status")
-		ar := h.onPending(ctx, cr)
-		err := h.client.CreateSubResource(cr.ResourceMeta, cr.SubResources)
+		credential = h.onPending(ctx, cr)
+	} else if actions := h.shouldProcessUpdating(cr); len(actions) != 0 {
+		log.Trace("processing resource in updating status")
+		credential = h.onUpdates(ctx, cr, actions)
+	}
+
+	if credential != nil {
+		err = h.client.CreateSubResource(cr.ResourceMeta, cr.SubResources)
 		if err != nil {
 			logger.WithError(err).Error("error creating subresources")
 		}
 
-		// update the status regardless of errors updating the other subresources
-		statusErr := h.client.CreateSubResource(ar.ResourceMeta, map[string]interface{}{"status": ar.Status})
+		// update the status resource regardless of errors updating the other subresources
+		statusErr := h.client.CreateSubResource(credential.ResourceMeta, map[string]interface{}{"status": credential.Status})
 		if statusErr != nil {
 			logger.WithError(statusErr).Error("error creating status subresources")
 			return statusErr
 		}
+	}
 
-		return err
+	return err
+}
+
+// shouldProcessDeleting
+func (h *credentials) shouldProcessDeleting(cr *management.Credential) bool {
+	if !hasAgentCredentialFinalizer(cr.Finalizers) {
+		return false
 	}
 
-	if ok := h.shouldProcessDeleting(cr.Status, cr.Metadata.State, cr.Finalizers); ok {
-		logger.Trace("processing resource in deleting state")
-		h.onDeleting(ctx, cr)
+	if cr.Spec.State.Name == v1.Inactive && cr.Spec.State.Reason == prov.CredExpDetail && cr.Status.Level == prov.Pending.String() {
+		// expired credential
+		return true
 	}
 
-	return nil
+	if cr.Metadata.State == v1.ResourceDeleting {
+		// don't process delete when error from agent
+		return !hasAgentCredentialError(cr.Status)
+	}
+
+	return false
 }
 
-func (h *credentials) getReasonMetaAction(reasons []apiv1.ResourceStatusReason) string {
-	if len(reasons) != 1 {
-		return ""
+// shouldProvision
+func (h *credentials) shouldProcessPending(cr *management.Credential) bool {
+	if h.marketplaceHandler.shouldProcessPending(cr.Status, cr.Metadata.State) {
+		return cr.Spec.State.Name == v1.Active && !cr.Spec.State.Rotate && cr.State.Name != v1.Inactive
 	}
-	if reasons[0].Meta == nil {
-		return ""
+	return false
+}
+
+// shouldProcessUpdating
+func (h *credentials) shouldProcessUpdating(cr *management.Credential) []prov.CredentialAction {
+	actions := []prov.CredentialAction{}
+	inter := reflect.TypeOf((*credProv)(nil)).Elem()
```

Comment:
what does this do?

Reply:
This is a check that the provisioner fully implements the Credential Interface. So we do not call update to an agent that does not have the update function implemented.